src/swaps/oneinch/OneinchSwapProvider.js

```diff
@@ -0,0 +1,263 @@
+    })
+    const toAmountInUnit = BN(trade.data.toTokenAmount)
+    return {
+      from,
+      to,
+      // TODO: Amounts should be in BigNumber to prevent loss of precision
+      fromAmount: fromAmountInUnit.toNumber(),
+      toAmount: toAmountInUnit.toNumber()
+    }
+  }
+
+  async approveTokens ({ network, walletId, quote }) {
+    const fromChain = cryptoassets[quote.from].chain
+    const toChain = cryptoassets[quote.to].chain
+    const chainId = ChainNetworks[fromChain][network].chainId
+    if (fromChain !== toChain || !chainToRpcProviders[chainId]) return null
+
+    const api = new ethers.providers.StaticJsonRpcProvider(chainToRpcProviders[chainId])
+    const erc20 = new ethers.Contract(cryptoassets[quote.from].contractAddress, ERC20.abi, api)
+    const fromAddressRaw = await this.getSwapAddress(network, walletId, quote.from, quote.toAccountId)
+    const fromAddress = chains[fromChain].formatAddress(fromAddressRaw)
+    const allowance = await erc20.allowance(fromAddress, this.routerAddress)
+    const inputAmount = ethers.BigNumber.from(BN(quote.fromAmount).toFixed())
+    if (allowance.gte(inputAmount)) {
+      return {
+        status: 'APPROVE_CONFIRMED'
+      }
+    }
+
+    const callData = await axios({
+      url: this.agent + `/${chainId}/approve/calldata`,
+      method: 'get',
+      params: { tokenAddress: cryptoassets[quote.from].contractAddress }
+    })
+
+    const account = this.getAccount(quote.fromAccountId)
+    const client = this.getClient(network, walletId, quote.from, account?.type)
+    const approveTx = await client.chain.sendTransaction(callData.data)
+
+    return {
+      status: 'WAITING_FOR_APPROVE_CONFIRMATIONS',
+      approveTx,
+      approveTxHash: approveTx.hash
+    }
+  }
+
+  async sendSwap ({ network, walletId, quote }) {
+    const toChain = cryptoassets[quote.to].chain
+    const fromChain = cryptoassets[quote.to].chain
+    const chainId = ChainNetworks[toChain][network].chainId
+    if (toChain !== fromChain || !chainToRpcProviders[chainId]) return null
+
+    const account = this.getAccount(quote.fromAccountId)
+    const client = this.getClient(network, walletId, quote.from, account?.type)
+    const fromAddressRaw = await this.getSwapAddress(network, walletId, quote.from, quote.fromAccountId)
+    const fromAddress = chains[toChain].formatAddress(fromAddressRaw)
+
+    const trade = await axios({
+      url: this.agent + `/${chainId}/swap`,
+      method: 'get',
+      params: { fromTokenAddress: assets[quote.from].contractAddress || nativeAssetAddress, toTokenAddress: assets[quote.to].contractAddress || nativeAssetAddress, amount: quote.fromAmount, fromAddress: fromAddress, slippage: 0.5 }
+    })
+    await this.sendLedgerNotification(quote, account, 'Signing required to complete the swap.')
```

Comment:
The quote displayed to the user from `/quote` call is different than the price executed later using `/swap`
So we should add an extra check here to compare `trade.toTokenAmount` to `quote.toAmount`. If it's larger than the slippage, throw an error

Reply:
Makes sense